application/models/user.go

```diff
@@ -239,6 +263,55 @@ func (u *User) FindOrCreateFromAuthUser(orgID int, authUser *auth.User) error {
 	return nil
 }
 
+// FindOrCreateFromOrglessAuthUser creates a new User based on an auth.User and
+// sets its SocialAuthProvider field so they can login again in future.
+func (u *User) FindOrCreateFromOrglessAuthUser(authUser *auth.User, authType string) error {
+
+	if err := DB.Where("email = ?", authUser.Email).First(u); err != nil {
+		if domain.IsOtherThanNoRows(err) {
+			return errors.WithStack(err)
+		}
+	}
+
+	newUser := true
+	if u.ID != 0 {
+		newUser = false
+	}
+
+	// update attributes from authUser
+	u.FirstName = authUser.FirstName
+	u.LastName = authUser.LastName
+	u.Email = authUser.Email
+	u.SocialAuthProvider = nulls.NewString(authType)
+
+	if authUser.PhotoURL != "" {
+		u.AuthPhotoURL = nulls.NewString(authUser.PhotoURL)
+	}
+
+	// if new user they will need a unique Nickname
+	if newUser {
+		u.Nickname = authUser.Nickname
+		if err := u.uniquifyNickname(); err != nil {
+			return err
+		}
+	}
+
+	if err := u.Save(); err != nil {
+		return fmt.Errorf("unable to save user record: %s", err.Error())
+	}
+
+	if newUser {
+		e := events.Event{
+			Kind:    domain.EventApiUserCreated,
+			Message: "Nickname: " + u.Nickname + "  UUID: " + u.UUID.String(),
+			Payload: events.Payload{"user": u},
+		}
+		emitEvent(e)
+	}
```

Comment:
Looks like 276-310 could be extracted, if not fully combining the two `FindOrCreate` methods.

Reply:
:+1: